### ISSUE TITLE
Possible bug: `acprop` explodes during initial steps

### ISSUE THREAD

**Issue by carlosgmartin** [google-deepmind/optax#1132]
There appears to be a bug in [`acprop`](https://optax.readthedocs.io/en/latest/api/contrib.html#optax.contrib.acprop) that causes it to explode during its initial steps in some situations:

<details>

<summary>Code</summary>

```python3
import jax
import optax
from jax import lax, numpy as jnp
from matplotlib import pyplot as plt, rcParams


def rosenbrock_function(params, a=1, b=100):
    x, y = params
    return jnp.square(a - x) + b * jnp.square(y - jnp.square(x))


def main():
    lr = 1e-4
    iters = 1000

    params = jnp.zeros(2)

    fig, ax = plt.subplots(constrained_layout=True)

    for label, opt in [
        ("sgd", optax.sgd(lr)),
        ("adam", optax.adam(lr)),
        ("adabelief", optax.adabelief(lr)),
        ("acprop", optax.contrib.acprop(lr)),
    ]:

        def update(state, _):
            params, opt_state = state
            value, grads = jax.value_and_grad(rosenbrock_function)(params)
            updates, opt_state = opt.update(grads, opt_state, params)
            params = optax.apply_updates(params, updates)
            return (params, opt_state), value

        opt_state = opt.init(params)
        _, values = lax.scan(update, (params, opt_state), length=iters)
        ax.plot(values, label=label)

    ax.legend()
    ax.set(xlabel="iteration", ylabel="value", yscale="log")
    ax.grid(True, "both", alpha=0.1)
    rcParams["savefig.dpi"] = 300
    plt.show()


if __name__ == "__main__":
    main()
```

</details>

<img src="https://github.com/user-attachments/assets/4fcbd75d-d678-44ce-a2ea-3bfaa933c11a" width="400">

Zooming in:

<img src="https://github.com/user-attachments/assets/e6198840-f968-4de1-90f8-6752bd77e1aa" width="400">

I'm currently investigating this.

**Comment by vroulet** [comment]
Hey @carlosgmartin,
Did you find the bug? We may report the bug to the original repo and author and add a note on the optimizer in the mean time (to be honest it's also fine if we don't add a note since many deep learning optimizers are quite frail anyway).
Once done we may close the issue.

**Comment by carlosgmartin** [comment]
@vroulet I didn't. @juntang-zhuang is one of the coauthors and has a repo [here](https://github.com/juntang-zhuang/ACProp-Optimizer). He might know what's causing the issue.

**Comment by vroulet** [comment]
I opened an issue on the author's repository [here](https://github.com/juntang-zhuang/ACProp-Optimizer/issues/2).
I'll close this one.

**Comment by aymuos15** [comment]
If still relevant, I was actually taking a look at this when I came across it during some personal lit survey.

First, a note on the `mu` difference: The prediction error in ACProp is `g - m` (gradient minus momentum). The original paper specifies `(g_t - m_t)²` using the *current* momentum, and the [PyTorch implementation](https://github.com/juntang-zhuang/ACProp-Optimizer/blob/main/optimizers/ACProp.py) follows this by computing `grad_residual = grad - exp_avg` after updating `exp_avg`. However, the [optax implementation](https://github.com/google-deepmind/optax/blob/main/optax/contrib/_acprop.py) uses `state.mu` (previous momentum `m_{t-1}`), which is inconsistent with both the paper and optax's own [docstring formula](https://github.com/google-deepmind/optax/blob/main/optax/contrib/_acprop.py#L122). That said, as shown below, this difference alone doesn't fix the convergence issue.

Another angle is the **`initial` state** handling. Both [PyTorch](https://github.com/juntang-zhuang/ACProp-Optimizer/blob/main/optimizers/ACProp.py#L161) (`step <= 1`) and [optax](https://github.com/google-deepmind/optax/blob/main/optax/contrib/_acprop.py#L72) (`count == 0`) treat 1 step as initial. As the graphs show below, have this as "2" shows convergence.

The final major difference: the PyTorch implementation has a `rectify` parameter (RAdam-style rectification) which is `True` by default but missing entirely from optax.

<img width="1524" height="1011" alt="Image" src="https://github.com/user-attachments/assets/7184f1b3-24f9-4fd0-8239-b260e8e74976" />

### What I tested

I created parameterized variants (which are equivalent to their original (jax and torch respectively)) testing these 3 factors:
1. **mu**: old (`state.mu`) vs new (`mu`) for prediction error
2. **initial**: 1 vs 2 steps treated as initial (before adaptive scaling)
3. **rectify**: with or without RAdam-style rectification

### Results (Rosenbrock function, lr=1e-4, 1000 iters)

| # | Config | Converges? |
|---|--------|------------|
| 1 | old mu, initial=1, no rectify | **NO** |
| 2 | new mu, initial=1, no rectify | **NO** |
| 3 | old mu, initial=2, no rectify | YES |
| 4 | new mu, initial=2, no rectify | YES |
| 5 | new mu, initial=2, rectify | YES |
| 6 | new mu, initial=1, rectify | **YES** |

**Key observations:**
- Variants 1 and 2 both fail → the `mu` fix alone doesn't help
- Variants 3-4 work → `initial=2` fixes convergence
- Variant 6 shows `rectify=True` works even with `initial=1`
- Variants 5-6 with `rectify=True` converge faster to a better final value, and this is what PyTorch uses by default


### The root cause (I am assuming based on only empirical results)

PyTorch works out of the box because `rectify=True` is the default. Optax explodes because it uses `initial=1` and has no rectification. Wanted to note that, I am not entirely sure what the best practices are here. Or even if setting the whole initial parameter to 2 is a valid approach so to say.

### Question

Should we then add a `rectify` parameter to optax's ACProp and follow the original PyTorch repo's defaults (`rectify=True`)?

---

Reference notebook with full comparison code: https://colab.research.google.com/drive/1f7ESiSAJ6EC5SDoK_dJFWAHCG1lmWVL2

**Comment by carlosgmartin** [comment]
@aymuos15 Thanks for the thorough investigation!

Might be worth contacting the [authors](https://arxiv.org/abs/2110.05454) to clarify the correct approach. A few of them appear to have GitHub accounts:
- https://github.com/juntang-zhuang
- https://github.com/crazyhottommy
- https://github.com/yifding

Unfortunately, the [repo](https://github.com/juntang-zhuang/ACProp-Optimizer) doesn't seem to be active.

**Comment by aymuos15** [comment]
Thanks for linking the accounts! I am guessing since there is already a request there, should be okay haha. Just wanted to add this on for future reference for someone else potentially!

**Comment by rdyro** [comment]
This looks like a really cool analysis, thank you both @carlosgmartin and @aymuos15!

Unless @vroulet objects, `new mu, initial=2, no rectify` seems like the most intuitive choice to me, @aymuos15  would you be willing to contribute a PR?

**Comment by aymuos15** [comment]
Thank you! Definitely happy to do a PR on this.
